software/controller/lib/core/breath_detection.cpp

```diff
@@ -0,0 +1,74 @@
+
+/// \TODO: This should be configurable from the GUI.
+static DebugFloat dbg_flow_trigger{"bd_flow_trigger", Debug::Variable::Access::ReadWrite, 200,
+                                   "mL/s", "Breath detection flow trigger"};
+
+// fast_alpha_ and slow_alpha_ were tuned for a control loop that runs at a particular frequency.
+// In theory if the control loop gets slower, the alpha terms should get bigger, placing more weight
+// on newer readings, and similarly if the control loop gets faster, the alpha terms should get
+// smaller.  We've tried to encode this here, although it remains to be seen if it actually works.
+static DebugFloat dbg_fast_alpha{"bd_fast_flow_avg_alpha", Debug::Variable::Access::ReadWrite,
+                                 0.2f * (Controller::GetLoopPeriod() / milliseconds(10)), "",
+                                 "alpha term for fast-updating exponentially-weighted average"
+                                 " of flow used for breath detection"};
+static DebugFloat dbg_slow_alpha{"bd_slow_flow_avg_alpha", Debug::Variable::Access::ReadWrite,
+                                 0.01f * (Controller::GetLoopPeriod() / milliseconds(10)), "",
+                                 "alpha term for slow-updating exponentially-weighted average"
+                                 " of flow used for breath detection"};
+static DebugFloat dbg_fast_flow_avg{"bd_fast_flow_avg", Debug::Variable::Access::ReadOnly, 0.0f,
+                                    "mL/s", "fast-updating flow average for breath detection"};
+static DebugFloat dbg_slow_flow_avg{"bd_slow_flow_avg", Debug::Variable::Access::ReadOnly, 0.0f,
+                                    "mL/s", "fast-updating flow average for breath detection"};
+
+BreathDetection::BreathDetection() {
+  dbg_fast_flow_avg.set(0);
+  dbg_slow_flow_avg.set(0);
+};
+
+bool BreathDetection::PatientInspiring(const BreathDetectionInputs &inputs, bool at_dwell) {
+  if (inputs.net_flow < ml_per_sec(0)) {
+    return false;
+  }
+
+  // Once flow is non-negative, start calculating two
+  // exponentially-weighted averages of net flow: slow_flow_avg_ and fast_flow_avg_.
+  //
+  // The slow one has a smaller alpha term, so updates slower than the fast one.
+  // You can think of the slow average as estimating "flow at dwell" and
+  // the fast average as estimating "current flow".
+  //
+  // If the fast average exceeds the slow average by a threshold, we trigger a breath.
+  float slow_alpha = dbg_slow_alpha.get();
+  float fast_alpha = dbg_fast_alpha.get();
+
+  slow_flow_avg_ = slow_alpha * inputs.net_flow + (1 - slow_alpha) * slow_flow_avg_;
+  fast_flow_avg_ = fast_alpha * inputs.net_flow + (1 - fast_alpha) * fast_flow_avg_;
+
+  dbg_slow_flow_avg.set(slow_flow_avg_.ml_per_sec());
+  dbg_fast_flow_avg.set(fast_flow_avg_.ml_per_sec());
+
+  return at_dwell && fast_flow_avg_ - slow_flow_avg_ > ml_per_sec(dbg_flow_trigger.get());
```

Comment:
Should the boolean evaluation of whether the patient is inspiring also be rendered visible via a debug variable?